File: 1802-number-of-students-unable-to-eat-lunch/number-of-students-unable-to-eat-lunch.py
from collections import deque

class Solution:
    def countStudents(self, students: List[int], sandwiches: List[int]) -> int:
        
        studentQueue = deque(students)
        sandwichesStack = sandwiches[::-1]

        sandwichMissed = 0

        while studentQueue:
            if sandwichesStack[-1] == studentQueue[0]:
                sandwichesStack.pop()
                studentQueue.popleft()
                sandwichMissed = 0
            else:
                studentQueue.append(studentQueue.popleft())
                sandwichMissed += 1

                if sandwichMissed == len(studentQueue):
                    break

        return len(studentQueue)










        """ count0 = students.count(0)
        count1 = students.count(1)

        for s in sandwiches:
            if s == 0:
                if count0 == 0:
                    return count1
                count0 -= 1
            else:
                if count1 == 0:
                    return count0
                count1 -= 1
        return 0 """

        

File: 1802-number-of-students-unable-to-eat-lunch/number-of-students-unable-to-eat-lunch.cpp
class Solution {
public:
    int countStudents(vector<int>& students, vector<int>& sandwiches) {
        int n = students.size();

        queue<int> studentQueue;
        stack<int> sandwichesStack;

        // Put students into queue
        for (int i = 0; i < n; i++) {
            studentQueue.push(students[i]);
        }

        // Put sandwiches into stack
        for (int i = n - 1; i >= 0; i--) {
            sandwichesStack.push(sandwiches[i]);
        }

        int sandwichMissed = 0;

        while (!studentQueue.empty()) {
            if (sandwichesStack.top() == studentQueue.front()) {
                sandwichesStack.pop();
                studentQueue.pop();
                sandwichMissed = 0;
            } else {
                studentQueue.push(studentQueue.front());
                studentQueue.pop();
                sandwichMissed++;

                if (sandwichMissed == studentQueue.size())
                    break;
            }
        }

        return studentQueue.size();
    }
};